kubeclient.gemspec

```diff
@@ -30,4 +30,5 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'rest-client'
   spec.add_dependency 'recursive-open-struct', '~> 1.0.4'
   spec.add_dependency 'http', '~> 2.2.2'
+  spec.add_dependency 'json-streamer'
```

Comment:
this library has 1 contributor and from my impression low activity/users/forks etc.
why was this specific one selected?

Reply:
From `json-streamer` description:

> The gem uses JSON::Stream's events in the background. It was chosen because it's a pure Ruby parser. A similar implementation can be done using the ~10 times faster Yajl::FFI gem that is dependent on the native YAJL library.

I think its quality is good enough, and considering an alternative of depending on a native library I'd much prefer using `json-streamer` 👍

Reply:
@moolitayer I don't know.  It hasn't been active for over 1 year.  Maybe it's stable or just not that widely used or both.  Has anyone reviewed the code?  It's very small so maybe if someone feels comfortable with the code, it would be ok to use.  Do we have a way to test it against a large sample set of json reponses to make us feel better about it.

Is this opt-in or a required change?  Will we bump versions to indicate this?

Reply:
json-streamer is literally [113 lines](https://github.com/thisismydesign/json-streamer/blob/master/lib/json/streamer/json_streamer.rb).  didn't follow in detail but my impression it's reasonably clean, has 100% test coverage (though tests don't use complicated data).  :+1:

- alternative: https://github.com/bdevel/json_stream_trigger — slightly more complicated, simplistic regexp-based JsonPath parser worries me, 8 commits 2 years ago and dormant since then.

json-stream: not very active but is used by 10 gems.  test look ...long :-) (though again "real world fixture" doc is 1 toy doc). :+1:

=> IMO this dep is fine for an opt-in feature.  It's not as battle-tested as JSON.parse, and slower, so it's good you changed the PR not to always use it.

Reply:
btw I think https://github.com/thisismydesign/json-streamer/commits/master was last updated in March this year

Reply:
Thanks, @moolitayer I must have misread it as 2016.  Yeah, I'm fine with this considering how small the project is and if it's opt-in.  Since rubygems doesn't have opt-in dependencies, we'd still have to bump the version if we add the new dependency.

Reply:
Can we make this an optional dependency, and raise a graceful error if this is attempted to be used and isn't available?  Haven't looked at the implementation in this PR recently if this is possible/viable, just a suggestion potentially remove some concerns regarding directly adding a new dependency to this gem.

That all said, going with this approach does then introduce complexity into the codebase for handling this, so it isn't something you would consider doing lightly.  Very much related to the discussion here:  https://github.com/abonas/kubeclient/issues/263

Reply:
If there are concerns, perhaps Kubeclient can start from offering streamed `as: :raw` interface, and leave it to caller (eg. ManageIQ) to use json-streamer (or whatever).

- Nick mentioned on #263 great example of less trivial use:
  Writing raw stream to temp file (to relieve k8s server ASAP), then take our time to stream-parse & process from that.

However, that way callers needs to know more about k8s response structure, error handling etc :-(
Exposing processing & error handling in reusable way would add a lot of "API surface"...
Whereas "we do everything, just give us a block" is neatly contained. :-)

@ilackarms personally I'm happy to review whatever direction you want to proceed (including current).

Reply:
@NickLaMuro we had a discussion about introducing optional dependencies back in https://github.com/abonas/kubeclient/pull/213. My view, given the comments above and that this gem does not bring additional dependencies we should be fine. @simon3z/@abonas have final call. (we would need to bump version, as mentioned)

Reply:
@cben personally i like the way the current patch is designed. one option i would consider is yielding `chunks` to the passed block if `as: :raw` option is specified, so the caller can handle parsing on its own. however I'm a big fan of decoupling; we should allow kubeclient to abstract the kube API as much as possible from callers, so they do not have to concern themselves with "low level" things like parsing JSON, specifying API endpoints, etc. just my $0.02

Reply:
@ilackarms @abonas @cben @jrafanie @kirs @moolitayer

Hi fellow developers,

I'm the creator and maintainer of the above mentioned [json-streamer](https://github.com/thisismydesign/json-streamer) gem.

Just wanted to let you know that I'm actively keeping an eye on my open source projects. This one did not require much attention lately hence the last release date is May 2017. I will take a look at the issue you raised. If it's reproducible it shouldn't take long to provide a fix.

Regarding features: IMO it fills a big hole that the currently available JSON parsers leave open and does that in a neat way. I found no similar implementation when I started but [json_stream_trigger](https://github.com/bdevel/json_stream_trigger) is actually a good catch. However it doesn't seem to be maintained. I will consider supporting JSONPath in the future.

Regarding community: I know quite a few use it and the number increases steadily since it's likely the best option out there. I'm not sure though how easy it is to find out about it. Since @dgraham [decided](https://github.com/dgraham/json-stream/issues/4#issue-224052592) not to mention it as an alternative to [json-stream](https://github.com/dgraham/json-stream/) it's not really advertised.

Regarding maintenance: I'm open to move the gem to an organization if you have any suggestions. For now all you have is my word. :)

Reply:
update: it looks like we will *not* be able to use `json-streamer` to do what we want here. it's currently possible to yield objects one-by-one from the `items` array in a kubernetes entity list, under the assumption that `items` is the only top-level array found in the JSON structure. it's not currently possible to specify the array by key name (see discussion at https://github.com/thisismydesign/json-streamer/issues/7#issuecomment-329816147).

thus we have two options: implement our own json-streamer type of gem, or accept the workaround which would break if Kubernetes change the JSON structure of *list objects.

i would be in favor of writing our own gem, but this will be additional tech debt for us to maintain

Reply:
Thought of a solution since: https://github.com/thisismydesign/json-streamer/issues/7#issuecomment-330254816

Reply:
@cben appears to be working now. updated! :+1: